ENH Multidimensional image base class

caspervdw opened the pull request:
This implements a base class that allows flexible reading and iteration through a file with multiple dimensions.  The properties `aggregate` and `iterate` define how each dimension functions, as discussed in #124 and #150 

I based the `ImageSequenceND` reader on this, I'll wait for your response until rebasing the `Bioformats` reader.

The base class itself is quite extensive, so that the derived classed can be slim:

    >>> class MDummy(Multidimensional):
    ...    @property
    ...    def pixel_type(self):
    ...        return 'uint8'
    ...    @property
    ...    def frame_shape_2D(self):
    ...        return self._frame_shape_2D
    ...    def __init__(self, shape, **dims):
    ...        for name in dims:
    ...            self.add_dim(name, dims[name])
    ...        self._frame_shape_2D = shape
    ...    def get_frame_2D(self, **ind):
    ...        return np.zeros(self.frame_shape_2D, dtype=self.pixel_type)

    >>> frames = MDummy((64, 64), t=80, c=2, z=10, m=5)
    >>> frames.aggregate = 'cz'
    >>> frames.iterate = 't'
    >>> frames.m.default = 3
    >>> frames[5]  # returns Frame at T=5, M=3 with shape (2, 10, 64, 64)

tacaswell commented:
Awesome.  I don't have time to read the code in detail right now, but it looks like you have sorted out how to solve some problems that I looked and gave up on as too hard :smile: .

danielballan commented on pims/base_frames.py:
```diff
@@ -528,3 +528,267 @@ def process(img_or_iterable, *args, **kwargs):
+
+    @property
+    def aggregate(self):
+        """ This determines which dimensions will be aggregated into one.
```
typo

danielballan commented:
Yes, this is good stuff. I will review carefully this week. I wonder if the multidimensional base class should become the only base class, making all of this stuff available by default.

danielballan commented on pims/base_frames.py:
```diff
@@ -528,3 +528,267 @@ def process(img_or_iterable, *args, **kwargs):
+            return []
+        for dim in self._dims:
+            if key == dim.name:
+                return dim
```
I think this approach fails silently if I try to access a dim that does not exist.

caspervdw replied:
That's right, is there a proper way to solve this? Just raise an `AttributeError`?

tacaswell replied:
Yes, that will let python fall back to `__getattribute__` and then if that fails bubble up to the user.

tacaswell replied:
these have been dealt with :heavy_check_mark:

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+
+    Parameters
+    ----------
+    name : string, dimension name
```
Please put description on new line other wise the numpydo parser will identify name as having the type 'string, dimension name'.

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+    ----------
+    name : string, dimension name
+    size : int, size of dimension
+    aggregate : boolean, default False
```
Things that are optional should be `name : type, optional`  with the default given in the docstring.

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
     return process
+
+
+class FramesDimension(object):
```
Might be worth looking at http://boltons.readthedocs.org/en/latest/namedutils.html#boltons.namedutils.namedlist which is mutable `namedtuple`. It will be lighter weight at runtime and less code for use to maintain.

tacaswell replied:
The other thing that occurs to me is we should be aware of https://github.com/xray/xray and see what we can steal from them.

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+    >>> frames.dims['m'].default = 3
+    >>> frames[5]  # returns Frame at T=5, M=3 with shape (2, 10, 64, 64)
+    """
+    def clear_dims(self):
```
Is this and `add_dim` really useful as public functions?  I have not mad it farther down the code yet, but it seems that once the reader is created, the dimensions should not change.

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+    def iterate(self, value):
+        for dim_iterate in value:
+            if dim_iterate.lower() in ['x', 'y']:
+                raise ValueError('Iterate does not take dimensions x or y.')
```
Induldging in a bit of bike shedding:

Why not `RuntimeError`?

caspervdw replied:
I guess I took `ValueError` because the user is trying to set a non-existing dimension to `iterate`

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+            if dim_iterate.lower() in ['x', 'y']:
+                raise ValueError('Iterate does not take dimensions x or y.')
+            if dim_iterate not in self.dims:
+                raise ValueError(("Dimension named '{}' does not exist " +
```
bikeshedding: why not `KeyError`

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+                raise ValueError(("Dimension named '{}' does not exist " +
+                                  "in this image.").format(dim_iterate))
+
+        new_dims_iter = [0] * len(value)
```
bikeshedding : use `None` for the place holder

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+        """
+        pass
+
+    def get_frame(self, i):
```
This is subtle as all get out

caspervdw replied:
Don't see exactly what you mean here, but I revised the code strongly so please take a look again at this part

tacaswell replied:
That was just a comment that this is a subtle problem and I think you got it more-or-less right (modulo where state is stored which is an implementation detail).

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+    default : int
+        when aggregate and iterate are both False, this is the fallback value
+    """
+    def __init__(self, name, size, aggregate=False, iterate=True, default=0):
```
can you add the attribute `i` to this class?

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+        # identify the indices to take along each dimension
+        i_prev = 0
+        for n, dim in enumerate(self._dims):
+            dim.i = dim.default
```
won't this do funny things if the default is set to 1/True?  My understanding from reading above was that this value was the default values for iterate/aggregate?

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+        iterate (together with the dimension default index) determines which.
+        multidimensional index is returned. """
+        # identify the indices to take along each dimension
+        i_prev = 0
```
I am not super wild about using an attribute on the dimensions to keep track of where you are, I think this code would be easier to understand as something like

```python
sizes = np.array([d.size for d in self.dims if d.iterate], dtype='int')
cumsizes = np.cumprod(sizes)
current_index = np.zeros(len(self._dims), dtype=int)
```

caspervdw replied:
Agreed, but this is a quite tricky part of the code. I now have this:

        # list sizes of iterate dimensions
        iter_sizes = [self._sizes[k] for k in self._iterate]
        # list how much i has to increase to get an increase of coordinate n
        iter_cumsizes = np.append(np.cumprod(iter_sizes[::-1])[-2::-1], 1)
        # calculate the coordinates and update the coords dictionary
        iter_coords = (i // iter_cumsizes) % iter_sizes

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+
+        # read all 2D frames
+        for n in range(Nframes):
+            result[n] = self.get_frame_2D(**{d.name: d.i for d in self._dims})
```
following the above:

```python
result[n] = self.get_frame_2d(**{d.name: indx for d, indx in zip(self._dims, cur_index)})
```

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+        Existing methods always precede over this: a dimension named 'ndim'
+        would not be accessible by frames.ndim. """
+        if key == '_dims':
+            return []
```
Is this to hide `_dims` or sholud this have been `return list(self._dims)`

caspervdw replied:
This was to properly initialize the _dims attribute without using __init__. But I moved it now to a check in `_dim_add`

tacaswell commented on pims/base_frames.py:
```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+        return s
+
+
+class Multidimensional(FramesSequence):
```
This should probably have an 

```python

def __init__(self, *args, **kwargs):
    self._dims = []
    super(Multidimensional, self).__init__(self, *arg, **kwargs)
```

caspervdw replied:
Ok, may be best solution

tacaswell commented on pims/image_sequence.py:
```diff
@@ -248,87 +246,84 @@ class ImageSequence3D(ImageSequence):
+        if as_grey:
+            raise ValueError('As grey not supported for ND images')
+        self.dim_identifiers = dim_identifiers
+        super(ImageSequenceND, self).__init__(path_spec, process_func,
```
do the super first if you can.

caspervdw replied:
Can't, the `__init__` requires the `dim_identifiers`

tacaswell commented on pims/image_sequence.py:
```diff
@@ -248,87 +246,84 @@ class ImageSequence3D(ImageSequence):
     def _get_files(self, path_spec):
-        super(ImageSequence3D, self)._get_files(path_spec)
-        self._toc = np.array([filename_to_tzc(f, self.tzc_identifiers) \
+        super(ImageSequenceND, self)._get_files(path_spec)
```
I have been working on a python3 only project, the better super support is great (`super().foo` works as expected)

tacaswell commented:
Left a whole bunch of comments, over all :+1: 

any comments where I used the phrase 'bike shedding' feel free to give minimal thought and dismiss.

Sorry this has sat for so long

caspervdw commented:
Thanks for the thorough review, I hope to give this a round sometime this
week.
Op 19 jun. 2015 19:04 schreef "Thomas A Caswell" <notifications@github.com>:

> Left a whole bunch of comments, over all [image: :+1:]
>
> any comments where I used the phrase 'bike shedding' feel free to give
> minimal thought and dismiss.
>
> Sorry this has sat for so long
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/soft-matter/pims/pull/151#issuecomment-113576867>.
>

caspervdw commented:
I refactored the code, tests pass (locally at least).

Inspired by xray I used the terminology _coordinate_ for an index along a specific dimension and _index_ for the value passed to `get_frame`. The transformation index --> coordinates according to `iterate` gave me some headaches and is (as far as I could see) not implemented in xray.

I kicked out the `FramesDimension` object and I just use dictionaries `_sizes` and `_default_coords` and lists `_aggrate` and `_iterate` to keep track of all the dimension properties. For clarity, the default coordinate is the coordinate that is used if the dimension is both not in `aggregate` and `iterate`

I didn't want to use the `__init__` because if you subclass for instance `ImageSequence` and `FramesSequenceND` together, you want to keep both `__init__` routines. Therefore I thought it to be more elegant to solve the dimension initialization in `_dim_add` if `_sizes` does not exist.

@tacaswell  I hope I covered all your suggestions, it's ready for another round of review and then I'll rebase and also base the Bioformats reader on this.

danielballan commented:
Slick. Thanks for tackling this subtle problem in a general way. Two comments:

1. I like the names borrowed from xray. The only name I don't like is aggregate, which to me suggested data reduction, as in SQL parlance or in pandas like: `df.groupby(...).agg(...)`. Are you open to using a different name? How about `bundle`?
2. Any `Frame.metadata` is ignored. This is the right choice for now -- better to drop it than propagate it badly -- but I just wanted to note it. Persisting metadata correctly and capturing more if it from the file formats we support can be a project for 0.3+.

caspervdw commented:
Rebased and added an extra update. Now `y` and `x` are treated equally as all dimensions. With some limitations due to the inherent 2D nature of image readers.

A thing left to do is handle interleaved formats (e.g. RGB)

caspervdw commented:
@danielballan I just implemented a metadata propagation support from `get_frame_2D` to `get_frame`. Metadata dicts of 2D frames get squashed in a dictionary of lists. If all values are the same in the list, it will be converted to a scalar. The derived reader may choose to include the coordinates of the 2D frames in the metadata.

I think there is no wrong metadata propagated in this way.

Separate thought: dictionaries aren't really the best format for storing multidimensional metadata.

danielballan commented:
OK, that sounds like a good approach to metadata. The general problem of propagating metadata through operations of Frame is much harder. (What happens when you add to Frames?) Pandas has raised this discussion and, for now, chosen to punt rather than guess and create some not-very-elegant conglomeration.

I like dictionaries for their flexible structure. What would you propose as an alternative? We should get this right going into 1.0.

Any comment on the nomenclature of "aggregate" versus an alternative like "bundle"? If you really like aggregate, I can be convinced, but I think both @tacaswell and I would prefer something that is not overloaded by existing tools.

caspervdw commented:
For me, "bundle" is ok. We can also do something more descriptive, such as "bundled_dims", or "bundled_axes".

What do you think of renaming "dim" to "axis", like in numpy?

As for the metadata, I guess the dictionaries are the best option for the general case. Just for the ND files, I would like to have some container that describes each plane (such as timestamp of each image), using multiindexing. A picture of size 10x2x512x512 should have a container of shape 10x2 that contains all the timestamps of each plane. Maybe an ndarray contained inside the dictionary is the best solution to that {'timestamps': ndarray}.

The way it is implemented now is as follows: {'z_index': list, 'c_index': list, 'timestamp': list}.

danielballan commented:
I like "axis." I usually go for verbose and descriptive, so bundled_axes
seems good to me. Bundled alone might confusingly suggest a Boolean.

I think an ndarray of metadata reflecting the structure of the bundled
Frame sounds right.
On Wed, Jun 24, 2015 at 4:58 AM Casper van der Wel <notifications@github.com>
wrote:

> For me, "bundle" is ok. We can also do something more descriptive, such as
> "bundled_dims", or "bundled_axes".
>
> What do you think of renaming "dim" to "axis", like in numpy?
>
> As for the metadata, I guess the dictionaries are the best option for the
> general case. Just for the ND files, I would like to have some container
> that describes each plane (such as timestamp of each image), using
> multiindexing. A picture of size 10x2x512x512 should have a container of
> shape 10x2 that contains all the timestamps of each plane. Maybe an ndarray
> contained inside the dictionary is the best solution to that {'timestamps':
> ndarray}.
>
> The way it is implemented now is as follows: {'z_index': list, 'c_index':
> list, 'timestamp': list}.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/soft-matter/pims/pull/151#issuecomment-114793042>.
>

caspervdw commented:
Ok, I renamed the following:
* `iterate` -> `iter_axes`
* `aggregate` -> `bundle_axes`
* `dims` -> `axes`
* `_add_dim` -> `_init_axis`

Also I added support and tests for metadata propagation in ndarrays.

danielballan commented:
OK, after one more careful look from us is this ready to merge on your end,
Casper?

On Wed, Jun 24, 2015 at 10:24 AM Casper van der Wel <
notifications@github.com> wrote:

> Ok, I renamed the following:
>
>    - iterate -> iter_axes
>    - aggregate -> bundle_axes
>    - dims -> axes
>    - _add_dim -> _init_axis
>
> Also I added support and tests for metadata propagation in ndarrays.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/soft-matter/pims/pull/151#issuecomment-114885410>.
>

danielballan commented on pims/image_sequence.py:
```diff
@@ -248,90 +246,86 @@ class ImageSequence3D(ImageSequence):
+        the same order as the order in this list. The last two elements have
+        to be ['y', 'x']. Defaults to ['z', 'y', 'x'].
+    default_coords: dict of int
+        When a dimension is not present in both iterate and aggregate, the
```
iterate and aggregate -> iter_axes and _bundle_axes

caspervdw commented:
It's ready for merge on my end, now also Bioformats is changed. Only thing is that the reader cannot return interleaved RGB or RGBA formats, because the last two axes are forced to ['y', 'x'].
But I don't have time to implement that until July 7th.

danielballan commented:
Looks like Travis is failing on bioformats:

```
AttributeError: 'BioformatsReader' object has no attribute '_source_dtype'
```

caspervdw commented:
Strange, I am sure tests passed locally... will have a look tomorrow!
Op 21 jun. 2015 10:11 schreef "Casper van der Wel" <caspervdw@gmail.com>:

> Thanks for the thorough review, I hope to give this a round sometime this
> week.
> Op 19 jun. 2015 19:04 schreef "Thomas A Caswell" <notifications@github.com
> >:
>
>> Left a whole bunch of comments, over all [image: :+1:]
>>
>> any comments where I used the phrase 'bike shedding' feel free to give
>> minimal thought and dismiss.
>>
>> Sorry this has sat for so long
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/soft-matter/pims/pull/151#issuecomment-113576867>.
>>
>

caspervdw commented:
@danielballan @tacaswell I fixed the tests, all should be fine now. I'll be away until July 7th, so I hope it can be merged. Otherwise, feel free to edit the code.

danielballan commented:
Great. Thanks for pushing this through before your departure.
